README.md
firewall Cookbook
=================

Provides a set of primitives for managing firewalls and associated rules.

PLEASE NOTE - The resource/providers in this cookbook are under heavy development. An attempt is being made to keep the resource simple/stupid by starting with less sophisticated firewall implementations first and refactor/vet the resource definition with each successive provider.


Requirements
------------
### Platform
* Ubuntu
* Debian

Tested on:
* Ubuntu 10.04
* Ubuntu 11.04
* Ubuntu 11.10
* Debian 7.0


Recipes
-------
### default
The default recipe installs the `ufw` package, which this cookbook requires. Make sure that the firewall recipe is in the node or role run_list before any resources from this cookbook is used.


Resources/Providers
-------------------
### firewall
#### Actions
- :enable: *Default action* enable the firewall.  this will make any rules that have been defined 'active'.
- :disable: disable the firewall. drop any rules and put the node in an unprotected state.

#### Attribute Parameters
- name: name attribute. arbitrary name to uniquely identify this resource
- log_level: level of verbosity the firewall should log at. valid values are: :low, :medium, :high, :full. default is :low.

#### Providers
- `Chef::Provider::FirewallUfw`
    - platform default: Ubuntu

#### Examples

```ruby
# enable platform default firewall
firewall 'ufw' do
  action :enable
end

# increase logging past default of 'low'
firewall 'debug firewalls' do
  log_level :high
  action    :enable
end
```

### firewall_rule

#### Actions
- :allow: the rule should allow incoming traffic.
- :deny: the rule should deny incoming traffic.
- :reject: *Default action: the rule should reject incoming traffic.

#### Attribute Parameters
- name: name attribute. arbitrary name to uniquely identify this firewall rule
- protocol: valid values are: :udp, :tcp. default is all protocols
- port: incoming port number (ie. 22 to allow inbound SSH)
- ports: array of incoming port numbers (ie. [80,443] to allow inbound HTTP & HTTPS). NOTE: `protocol` attribute is required with `ports`
- port_range: range of incoming port numbers (ie. 60000..61000 to allow inbound mobile-shell. NOTE: `protocol` attribute is required with `port_range`
- source: ip address or subnet to filter on incoming traffic. default is `0.0.0.0/0` (ie Anywhere)
- destination: ip address or subnet to filter on outgoing traffic.
- dest_port: outgoing port number.
- position: position to insert rule at. if not provided rule is inserted at the end of the rule list.
- direction: direction of the rule. valid values are: :in, :out, default is :in
- interface: interface to apply rule (ie. 'eth0').
- logging: may be added to enable logging for a particular rule. valid values are: :connections, :packets. In the ufw provider, :connections logs new connections while :packets logs all packets.

#### Providers

- `Chef::Provider::FirewallRuleUfw`
    - platform default: Ubuntu

#### Examples

```ruby
# open standard ssh port, enable firewall
firewall_rule 'ssh' do
  port     22
  action   :allow
  notifies :enable, 'firewall[ufw]'
end

# open standard http port to tcp traffic only; insert as first rule
firewall_rule 'http' do
  port     80
  protocol :tcp
  position 1
  action   :allow
end

# restrict port 13579 to 10.0.111.0/24 on eth0
firewall_rule 'myapplication' do
  port      13579
  source    '10.0.111.0/24'
  direction :in
  interface 'eth0'
  action    :allow
end

# open UDP ports 60000..61000 for mobile shell (mosh.mit.edu), note
# that the protocol attribute is required when using port_range
firewall_rule 'mosh' do
  protocol   :udp
  port_range 60000..61000
  action     :allow
end

# open multiple ports for http/https, note that the protocol
# attribute is required when using ports
firewall_rule 'http/https' do
  protocol :tcp
  ports    [80, 443]
  action   :allow
end

firewall 'ufw' do
  action :nothing
end
```


Development
-----------
This section details "quick development" steps. For a detailed explanation, see [[Contributing.md]].

1. Clone this repository from GitHub:

        $ git clone git@github.com:opscode-cookbooks/firewall.git

2. Create a git branch

        $ git checkout -b my_bug_fix

3. Install dependencies:

        $ bundle install

4. Make your changes/patches/fixes, committing appropiately
5. **Write tests**
6. Run the tests:
    - `bundle exec foodcritic -f any .`
    - `bundle exec rspec`
    - `bundle exec rubocop`
    - `bundle exec kitchen test`

  In detail:
    - Foodcritic will catch any Chef-specific style errors
    - RSpec will run the unit tests
    - Rubocop will check for Ruby-specific style errors
    - Test Kitchen will run and converge the recipes


License 
--------
```text
Copyright:: Copyright (c) 2015, Cloudenablers.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
firewall Cookbook CHANGELOG
=======================
This file is used to list changes made in each version of the firewall cookbook.


v0.11.8 (2014-05-20)
--------------------
* Corrects issue where on a secondary converge would not distinguish between inbound and outbound rules


providers/ufw.rb
#
# Cookbook Name:: firwall
# Provider:: ufw
#
# Copyright:: 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include Chef::Mixin::ShellOut

action :enable do
  if active?
    Chef::Log.debug("#{@new_resource} already enabled.")
  else
    shell_out!('echo yes | ufw enable')
    Chef::Log.info("#{@new_resource} enabled")
    if @new_resource.log_level
      shell_out!("ufw logging #{@new_resource.log_level}")
      Chef::Log.info("#{@new_resource} logging enabled at '#{@new_resource.log_level}' level")
    end
    new_resource.updated_by_last_action(true)
  end
end

action :disable do
  if active?
    shell_out!('ufw disable')
    Chef::Log.info("#{@new_resource} disabled")
    new_resource.updated_by_last_action(true)
  else
    Chef::Log.debug("#{@new_resource} already disabled.")
  end
end

private

def active?
  @active ||= begin
    cmd = shell_out!('ufw status')
    cmd.stdout =~ /^Status:\sactive/
  end
end


resources/rule.rb
#
# Cookbook Name:: firwall
# Resource:: rule
#
# Copyright:: 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

IP_CIDR_VALID_REGEX = /\b(?:\d{1,3}\.){3}\d{1,3}\b(\/[0-3]?[0-9])?/

actions :allow, :deny, :reject

attribute :port, :kind_of => Integer
attribute :ports, :kind_of => Array
attribute :port_range, :kind_of => Range
attribute :protocol, :kind_of => Symbol, :equal_to => [:udp, :tcp]
attribute :direction, :kind_of => Symbol, :equal_to => [:in, :out]
attribute :interface, :kind_of => String
attribute :logging, :kind_of => Symbol, :equal_to => [:connections, :packets]
attribute :source, :regex => IP_CIDR_VALID_REGEX
attribute :destination, :regex => IP_CIDR_VALID_REGEX
attribute :dest_port, :kind_of => Integer
attribute :position, :kind_of => Integer

def initialize(name, run_context = nil)
  super
  set_platform_default_providers
  @action = :reject
end

private

def set_platform_default_providers
  [:ubuntu, :debian].each do |platform|
    Chef::Platform.set(
        :platform => platform,
        :resource => :firewall_rule,
        :provider => Chef::Provider::FirewallRuleUfw
    )
  end
end


resources/default.rb
#
# Cookbook Name:: firwall
# Resource:: default
#
# Copyright:: 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

actions :enable, :disable

attribute :log_level, :kind_of => Symbol, :equal_to => [:low, :medium, :high, :full], :default => :low

def initialize(name, run_context = nil)
  super
  set_platform_default_providers
  @action = :enable
end

private

def set_platform_default_providers
  [:ubuntu, :debian].each do |platform|
    Chef::Platform.set(
      :platform => platform,
      :resource => :firewall,
      :provider => Chef::Provider::FirewallUfw
    )
  end
end


providers/rule_ufw.rb
#
# Cookbook Name:: firwall
# Provider:: rule_ufw
#
# Copyright:: 2015, Cloudenablers
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

def whyrun_supported?
  true
end

include Chef::Mixin::ShellOut

action :allow do
  if rule_exists?
    Chef::Log.debug "Rule #{rule} already allowed - skipping"
  else
    converge_by("Allowing #{rule}") do
      apply_rule('allow')
    end
  end
end

action :deny do
  if rule_exists?
    Chef::Log.debug "Rule #{rule} already denied - skipping"
  else
    converge_by("Denying #{rule}") do
      apply_rule('deny')
    end
  end
end

action :reject do
  if rule_exists?
    Chef::Log.debug "Rule #{rule} already rejected - skipping"
  else
    converge_by("Rejecting #{rule}") do
      apply_rule('reject')
    end
  end
end

private

# ufw allow from 192.168.0.4 to any port 22
# ufw deny proto tcp from 10.0.0.0/8 to 192.168.0.1 port 25
# ufw insert 1 allow proto tcp from 0.0.0.0/0 to 192.168.0.1 port 25
def apply_rule(type = nil) # rubocop:disable MethodLength
  ufw_command = 'ufw '
  ufw_command << "insert #{@new_resource.position} " if @new_resource.position
  ufw_command << "#{type} "
  ufw_command << "#{rule} "

  Chef::Log.debug("ufw: #{ufw_command}")
  shell_out!(ufw_command)

  Chef::Log.info("#{@new_resource} #{type} rule added")
  shell_out!('ufw status verbose') # purely for the Chef::Log.debug output
  new_resource.updated_by_last_action(true)
end

def logging
  case @new_resource.logging
  when :connections
    'log '
  when :packets
    'log-all '
  else
    ''
  end
end

def rule
  rule = ''
  rule << "#{@new_resource.direction} " if @new_resource.direction
  if @new_resource.interface
    if @new_resource.direction
      rule << "on #{@new_resource.interface} "
    else
      rule << "in on #{@new_resource.interface} "
    end
  end
  rule << logging
  rule << "proto #{@new_resource.protocol} " if @new_resource.protocol
  if @new_resource.source
    rule << "from #{@new_resource.source} "
  else
    rule << 'from any '
  end
  rule << "port #{@new_resource.dest_port} " if @new_resource.dest_port
  if @new_resource.destination
    rule << "to #{@new_resource.destination} "
  else
    rule << 'to any '
  end
  if @new_resource.port
    rule << "port #{@new_resource.port} "
  elsif @new_resource.ports
    rule << "port #{@new_resource.ports.join(',')} "
  elsif @new_resource.port_range
    rule << "port #{@new_resource.port_range.first}:#{@new_resource.port_range.last} "
  end
  rule.strip
end

# TODO: currently only works when firewall is enabled
def rule_exists?
  # To                         Action      From
  # --                         ------      ----
  # 22                         ALLOW       Anywhere
  # 192.168.0.1 25/tcp         DENY        10.0.0.0/8
  # 22                         ALLOW       Anywhere
  # 3309 on eth9               ALLOW       Anywhere
  # Anywhere                   ALLOW       Anywhere
  # 80                         ALLOW       Anywhere (log)
  # 8080                       DENY        192.168.1.0/24
  # 1.2.3.5 5469/udp           ALLOW       1.2.3.4 5469/udp
  # 3308                       ALLOW       OUT Anywhere on eth8

  to = ''
  to << "#{Regexp.escape(@new_resource.destination)}\s" if @new_resource.destination

  if @new_resource.protocol && @new_resource.port
    to << "#{Regexp.escape("#{@new_resource.port}/#{@new_resource.protocol}")}\s"
  elsif @new_resource.port
    to << "#{Regexp.escape("#{@new_resource.port}")}\s"
  end

  to << "Anywhere\s" if to.empty?

  action = @new_resource.action
  action = action.first if action.is_a?(Enumerable)
  action = "#{Regexp.escape(action.to_s.upcase)}\s"

  from = ''
  from << "#{Regexp.escape(@new_resource.source || 'Anywhere')}"

  if @new_resource.direction == :out
    regex = /^#{to}.*#{action}OUT\s.*#{from}.*$/
  else
    regex = /^#{to}.*#{action}.*#{from}.*$/
  end

  match = shell_out!('ufw status').stdout.lines.find do |line|
    # TODO: support IPv6
    return false if line =~ /\(v6\)$/
    line =~ regex
  end

  Chef::Log.debug("ufw: found existing rule for \"#{rule}\": \"#{match.strip}\"") if match

  !!match
end
